annotatedcsv/decode.go

```diff
@@ -0,0 +1,451 @@
+		fieldsMap := make(map[string]reflect.StructField)
+		err := forEachField(et, func(f reflect.StructField, name string) error {
+			fieldsMap[name] = f
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+
+		f = func(col Column) reflect.Type {
+			if f, ok := fieldsMap[col.Name]; ok {
+				return f.Type
+			}
+			return nil
+		}
+	case reflect.Slice:
+		s := et.Elem()
+		if s.Kind() != reflect.String && s.Kind() != reflect.Interface {
+			return fmt.Errorf("cannot decode into pointer to slice of %v", s)
+		}
+		f = func(col Column) reflect.Type {
+			return s
+		}
+	}
+	setters, err := fieldSetters(columns, f)
+	if err != nil {
+		return err
+	}
+	r.colSetters = setters
+	r.decodeType = t
+
+	return nil
+}
+
+// convertColumnValue set a value from current row of given column index
+// to a struct or a slice field value
+func (r *Reader) convertColumnValue(v reflect.Value, colIndex int) error {
+	s := stringTernary(r.row[colIndex], r.cols[colIndex].Default)
+	if err := r.colSetters[colIndex](v, s); err != nil {
+		return fmt.Errorf(`cannot convert value "%s" to type "%s" at line %d: %w`, s, r.cols[colIndex].Type, r.r.Line(), err)
+	}
+	return nil
+}
+
+type fieldFunc func(f reflect.StructField, name string) error
+
+// forEachField enumerates visible fields of t, finds field name and calls field function
+func forEachField(t reflect.Type, ff fieldFunc) error {
+	fields := ireflect.VisibleFields(t)
+	for _, f := range fields {
+		name := f.Name
+		if tag, ok := f.Tag.Lookup("flux"); ok {
+			if tag == "-" {
+				continue
+			}
+			name = tag
+		}
+		if err := ff(f, name); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// fieldKind represents one of the possible kinds of struct field.
+// It's similar to reflect.Kind (every reflect.Kind constant i
+// is represented as fieldKind(i)) except that it also has
+// defined constants for field types that have special treatment,
+// such as time.Duration.
+type fieldKind uint
+
+// conv represents a possible conversion source and destination.
+type conv struct {
+	from colType
+	to   fieldKind
+}
+
+// intKinds holds all supported signed integer kinds
+var intKinds = []fieldKind{
+	fieldKind(reflect.Int),
+	fieldKind(reflect.Int8),
+	fieldKind(reflect.Int16),
+	fieldKind(reflect.Int32),
+	fieldKind(reflect.Int64),
+}
+
+// uintKinds holds all supported unsigned integer kinds
+var uintKinds = []fieldKind{
+	fieldKind(reflect.Uint),
+	fieldKind(reflect.Uint8),
+	fieldKind(reflect.Uint16),
+	fieldKind(reflect.Uint32),
+	fieldKind(reflect.Uint64),
+}
+
+// floatKinds holds all supported floating point number kinds
+var floatKinds = []fieldKind{
+	fieldKind(reflect.Float32),
+	fieldKind(reflect.Float64),
+}
+
+// rest of supported type kinds
+const (
+	durationKind = fieldKind(255 + iota)
+	timeKind
+	bytesKind
+)
+
+// colType is integer based representation annotated CSV types
+type colType uint
+
+const (
+	stringCol = colType(iota)
+	boolCol
+	durationCol
+	longCol
+	uLongCol
```

Comment:
Let's make this reflect the original naming, even if it is a tad verbose:

```suggestion
	unsignedLongCol
```

Reply:
Ok